yarn-project/noir-libs/value-note/src/utils.nr

```diff
@@ -1,184 +1,125 @@
-
-    // Ensure the notes are actually owned by the owner (to prevent user from generating a valid proof while
-    // spending someone else's notes).
-    assert(owner == note.owner);
-
-    // Removes the note from the owner's set of notes.
-    balance.remove(context, note);
-
-    let note_value = note.value;
-    
-    // Assert that the note chosen to spend has enough funds.
-    assert(note_value as u64 >= amount as u64);
-
-    // Creates change note for the owner.
-    let change_value = note_value - amount;
-    let mut change_note = ValueNote::new(change_value, owner);
-
-    // Insert the change note to the owner's sets of notes.
-    balance.insert(context, &mut change_note);
-
-    // Emit the newly created encrypted note preimages via oracle calls.
-    let mut encrypted_data = [0; VALUE_NOTE_LEN];
-    if change_value != 0 {
-        encrypted_data = change_note.serialise();
-    };
-
-    let encryption_pub_key = get_public_key(owner);
-    emit_encrypted_log(
-        context,
-        context.inputs.call_context.storage_contract_address,
-        balance.storage_slot,
-        encryption_pub_key,
-        encrypted_data,
-    );
+    let sum = decrease_balance_max_amount(context, balance, amount, owner);
+    assert(sum == amount);
 }
 
-fn send_note(
+// Similar to `decrease_balance`, except that it doesn't fail if the subtracted amount is less than max_amount.
+// It returns the subtracted value, which should be less than or equal to max_amount.
+fn decrease_balance_max_amount(
     context: &mut PrivateContext,
     balance: Set<ValueNote, VALUE_NOTE_LEN>,
-    amount: Field,
-    recipient: Field,
-) {
-     // Creates new note for the recipient.
-    let mut note = ValueNote::new(amount, recipient);
-
-    // Insert the new note to the recipient's set of notes.
-    balance.insert(context, &mut note);
-
-    // Emit the newly created encrypted note preimages via oracle calls.
-    // docs:start:encrypted
+    max_amount: Field,
+    owner: Field,
+) -> Field {
+    let options = create_note_getter_options_for_decreasing_balance(max_amount);
+    let opt_notes = balance.get_notes(context, options);
 
-    let application_contract_address = (*context).this_address();
-    let note_storage_slot = balance.storage_slot;
-    let encryption_pub_key = get_public_key(recipient);
-    let encrypted_data = note.serialise();
+    let mut subtracted = 0;
+    for i in 0..opt_notes.len() {
+        if opt_notes[i].is_some() {
+            subtracted += destroy_note(context, balance, owner, opt_notes[i].unwrap_unchecked());
+        }
+    }
 
-    emit_encrypted_log(
-        context,
-        application_contract_address,
-        note_storage_slot,
-        encryption_pub_key,
-        encrypted_data,
-    );
+    // Add the change value back to the owner's balance.
+    let mut change_value = 0;
+    if subtracted as u120 > max_amount as u120 {
+        change_value = subtracted - max_amount;
+        subtracted -= change_value;
```

Comment:
These operations should probably be applied to `u120` types instead, to prevent underflow (if `u120` is indeed protected from underflows?)

Reply:
Scratch that. Lasse wrote a type safe u120 library: https://github.com/AztecProtocol/aztec-packages/issues/1794